#635 - Add pref experiment type checks

andymikulski opened the pull request:
Fixes #635 

- Adds type checks when setting preference experiment values

---

I'm having trouble getting the test(s) for this feature to work. It doesn't seem to catch errors as it should, and manually throwing hard errors in `start` doesn't bubble up to the test runner, either.  I'm wondering if I have the `yield` syntax incorrect or something.

@Osmose @mythmon If you could offer some feedback on this, that'd be super appreciated!

Osmose commented on recipe-client-addon/lib/PreferenceExperiments.jsm:
```diff
@@ -142,6 +142,16 @@ this.PreferenceExperiments = {
 
+    if (experiment.previousPreferenceValue) {
+      const previousType = typeof experiment.previousPreferenceValue;
+      const givenType = typeof preferenceValue;
```
Instead of using `typeof`, there's an API for getting the type of a preference that you'll want to use: [nsiPrefBranch.getPrefType](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsIPrefBranch#getPrefType()).

That interface is available on `Services.prefs`. `RecipeRunner.jsm` is an example of how to import and use that object.

andymikulski replied:
@Osmose It doesn't seem that I can just give the method a value and have it spit the type of out; it takes a pref name and checks the current value of that pref.

Does that mean that in order to test the new given value, it needs to first set the pref, then check the type, then revert the value if necessary?

I looked through the docs and didn't see anything that would help cut out this weird set-then-reset business, but maybe I missed something. Thoughts?

Osmose replied:
> @Osmose It doesn't seem that I can just give the method a value and have it spit the type of out; it takes a pref name and checks the current value of that pref.

Are you sure? I just ran this in my browser console and it worked fine:

```js
>> Services.prefs.getPrefType("asdf") // Boolean pref
128
>> Services.prefs.getPrefType("asdf") === Services.prefs.PREF_BOOL
true
>> Services.prefs.getPrefType("asdfasdf") // Doesn't exist
0
>> Services.prefs.getPrefType("asdfasdf") === Services.prefs.PREF_INVALID
true
```

Note that it's returning ints. The doc page mentioned that the return values were `PREF_BOOL` and the like, which I found as constants on the `Services.prefs` object. Sorry for not checking and explaining that earlier.

Osmose replied:
Also, the docs for `getPrefType` doesn't mention it, but the `PREF_BOOL` return values are constants defined on `Services.prefs`. So `Services.prefs.PREF_BOOL` can be used to check if the preference is a boolean, and so forth.

andymikulski replied:
Alright, I've corrected the approach and use the strings we use in the recipe arguments accordingly. _However,_ I'm having trouble figuring out how to actually read the recipe's arguments from within `PreferenceExperiments.jsm`. I've found that the RecipeRunner has access to the `recipe` object, but I couldn't figure out how to pass that into the PrefExperiments file.

Any suggestions?

mythmon commented:
> I'm having trouble getting the test(s) for this feature to work. It doesn't seem to catch errors as it should, and manually throwing hard errors in start doesn't bubble up to the test runner, either. I'm wondering if I have the yield syntax incorrect or something.

I ran into this too, and Osmose and I tracked it down: it turns out that one of our `with...` helper methods doesn't work with generators, and only works with async/await functions. Converting the test to async/await made them work for me. It was as simple as replace `yield` with `await` and `function*` with `async function`.

andymikulski commented:
I was having a real hard time getting tests to run/fail accordingly, but I just rebased my branch on `mozilla/preference-experiment` and it all seems to be working properly now! Wish I had done that sooner.. This should be ready for review shortly.

andymikulski commented:
@Osmose can you take a look at 9837a97541206066534a4595f7020230fcf32aed and tell me if something's missing? The `start` test isn't receiving the `preferenceType` property I've added, but I'm not sure why. Thoughts?

Osmose commented:
> @Osmose can you take a look at 9837a97 and tell me if something's missing? The start test isn't receiving the preferenceType property I've added, but I'm not sure why. Thoughts?

The test that you added is actually working fine. When I run the tests locally, I see the error:

```
 0:01.58 LOG: Thread-9 ERROR Unexpected exception Error: Previous preference value is of type "32", but was given "0" (undefined) at resource://shield-recipe-client/lib/PreferenceExperiments.jsm:217
start@resource://shield-recipe-client/lib/PreferenceExperiments.jsm:217:15
async*@/Users/mkelly/Projects/mozilla-unified/obj-x86_64-apple-darwin16.4.0/_tests/xpcshell/browser/extensions/shield-recipe-client/test/unit/test_PreferenceExperiments.js:88:9
async*inner@utils.js:15:35
async*inner@resource://shield-recipe-client/lib/PreferenceExperiments.jsm:131:37
async*asyncFunction@resource://gre/modules/Task.jsm:241:18
Task_spawn@resource://gre/modules/Task.jsm:166:12
_run_next_test@/Users/mkelly/Projects/mozilla-unified/testing/xpcshell/head.js:1554:9
run@/Users/mkelly/Projects/mozilla-unified/testing/xpcshell/head.js:703:9
_do_main@/Users/mkelly/Projects/mozilla-unified/testing/xpcshell/head.js:211:5
_execute_test@/Users/mkelly/Projects/mozilla-unified/testing/xpcshell/head.js:546:5
@-e:1:1
```

Note the `test_PreferenceExperiments.jsm` line; it says line 88 is where it failed, which isn't inside the new test you added. It's in another test for `start`, which is not passing the new `preferenceType` parameter. You'll have to update the existing tests to pass the new parameter as well as adding a new test.

Also, there's a bug in `PreferenceExperiments.jsm` in that commit, on line 212 you're indexing into the `typeConversions` object using `givenType`, when the index should be `argumentValueType`.

Also, we just recently found out that the tests in `test_PreferenceExperiments.jsm` are broken because `PreferenceExperiments.jsm` now relies on Telemetry, which isn't compatible with xpcshell tests. Even if you fix these issues, your test runs should fail. @mythmon is, as we speak, working on a patch to convert the test suite to mochitests. In the meantime, if you can get the code working in a manual test, I can do a code review, until the test fix is done.

andymikulski commented:
Locally, I get completely different test results:

```
 0:00.96 LOG: Thread-9 ERROR Unexpected exception Error: Previous preference value is of type "32", but was given "0" (undefined) at resource://shield-recipe-client/lib/PreferenceExperiments.jsm:217
start@resource://shield-recipe-client/lib/PreferenceExperiments.jsm:217:15
_do_main@/Users/amikulski/ff-src/mozilla-central/testing/xpcshell/head.js:211:5
_execute_test@/Users/amikulski/ff-src/mozilla-central/testing/xpcshell/head.js:546:5
@-e:1:1
```

I will make the corrections you've mentioned in the above post, however I think I'm going to have to wipe everything and see if a clean 'install' fixes my issues.

Are there any docs on working with the `recipe-client-addon`? I've never actually seen anything that details how to get tests running etc.

Osmose commented:
> Are there any docs on working with the recipe-client-addon? I've never actually seen anything that details how to get tests running etc.

Looks like I'll be writing docs today. Which is to say, "No, and there should be."

andymikulski commented:
Wiping my mozilla-central and re-cloning seems to have fixed my issues. I should be able to effectively patch the remainder of this bug now!

andymikulski commented:
I've patched the remaining tests for pref type checking, however there appears to be an unrelated test failure (that even exists on the actual `preference-experiment` branch). Specifically, it relates to the `start did not modify the default preference` test - I don't know if this is a known issue or not.

Either way, this PR should be ready for review! @Osmose @mythmon

mythmon commented:
Since we need a Firefox peer to review changes to the add-on, I'm going to tag in @gijsk instead of me to review this.

gijsk commented on recipe-client-addon/lib/PreferenceExperiments.jsm:
```diff
@@ -186,10 +187,39 @@ this.PreferenceExperiments = {
+      // Default to an INVALID pref type, unless there is a key in the typeConversions
+      // that matches the given argument type
+      let givenType = Services.prefs.PREF_INVALID || 0;
+      if(typeConversions.hasOwnProperty(argumentValueType)){
```
Nit: spaces before `(` and after `)`

This should fail eslint (./mach eslint path/to/file inside m-c). Does it not?

andymikulski replied:
It does! I had been hitting an issue with eslint not running properly, but after some fiddling it's picking this up appropriately now.

gijsk commented on recipe-client-addon/lib/PreferenceExperiments.jsm:
```diff
@@ -186,10 +187,39 @@ this.PreferenceExperiments = {
+
+      // Default to an INVALID pref type, unless there is a key in the typeConversions
+      // that matches the given argument type
+      let givenType = Services.prefs.PREF_INVALID || 0;
```
Shouldn't need to `||` this with 0.

gijsk commented on recipe-client-addon/test/unit/test_PreferenceExperiments.js:
```diff
@@ -165,6 +172,24 @@ add_task(withMockExperiments(withMockPreferences(async function (experiments, mo
+  mockPreferences.set("fake.type_preference", "oldvalue");
+
+  await Assert.rejects(
+    PreferenceExperiments.start({
```
We should also have a test that the forwarded (wrapAsync) stuff on the NormandyDriver throws, too (so we don't accidentally swallow exceptions there).

More generally, we should make sure that the start() callers do the right thing with these exceptions. I tried to find them but github search is awful and this all seems to still live on a branch, maybe / sort of? So I couldn't figure out where the entrypoint is... 

In any case, we should make sure that if starting one of these experiments fails, it doesn't break running any/all experiments. Especially because, for preferences that don't live in modules/libpref/init/all.js and friends (ie appear in about:config on a default profile), it's possible for users to define those prefs with an unexpected type, and we wouldn't want users doing that to be breaking their browsers and/or all our code.

Osmose replied:
> We should also have a test that the forwarded (wrapAsync) stuff on the NormandyDriver throws, too (so we don't accidentally swallow exceptions there).

That's probably a test for a different PR, in the SandboxManager tests. I'd rather test that once for wrapAsync instead of testing every use of wrapAsync. Although there is the unfilled gap of an integration test for preference-experiment recipes.

> More generally, we should make sure that the start() callers do the right thing with these exceptions. I tried to find them but github search is awful and this all seems to still live on a branch, maybe / sort of? So I couldn't figure out where the entrypoint is...

The entrypoint is in the action code in `recipe-server` that gets run in the sandbox: https://github.com/andymikulski/normandy/blob/d74e71bc7c86d392e193d79a3cc819bb7d2427c6/recipe-server/client/actions/preference-experiment/index.js#L40

Feature branches are... well, they are _a_ solution to a real problem. It's still up for debate if they're a _good_ solution. :P

> In any case, we should make sure that if starting one of these experiments fails, it doesn't break running any/all experiments.

We have existing tests that [ActionSandboxManager throws when the sandbox code throws](https://github.com/andymikulski/normandy/blob/d74e71bc7c86d392e193d79a3cc819bb7d2427c6/recipe-client-addon/test/unit/test_ActionSandboxManager.js#L70). We're missing a test that `RecipeRunner`, which calls that method, handles the error appropriately and doesn't halt execution of other actions. I think that's separate from this PR though, and shouldn't block it.

gijsk commented on recipe-client-addon/lib/PreferenceExperiments.jsm:
```diff
@@ -186,10 +187,39 @@ this.PreferenceExperiments = {
 
+    // If there's a previous preference value set, we need to ensure that the
+    // incoming value is of the same type.
+    if (preferences.has(preferenceName)) {
```
My understanding is that if:

1) a pref experiment uses a user-branch pref
2) there's a default branch value

this code won't detect it. Is that intentional? (I don't think it should be, but it's possible I'm missing something...)

Osmose commented on recipe-client-addon/test/unit/test_PreferenceExperiments.js:
```diff
@@ -148,11 +155,11 @@ add_task(withMockExperiments(withMockPreferences(async function (experiments, mo
   };
-  const experiment = {};
-  Object.keys(expectedExperiment).forEach(key => experiment[key] = experiments.test[key]);
+  const experiment = Object.assign({}, expectedExperiment);
```
The original lines copied `experiments.test`, but only copied the keys from `experiment`. The intent was to only do a comparison of the values we care about, in case the experiment data ended up including more data later. The new line doesn't handle that, and should be changed back.

Although to be honest it's a confusing bit of code, and I should've left a comment mentioning that in the first place. My bad!

Osmose commented on recipe-server/client/actions/preference-experiment/index.js:
```diff
@@ -43,6 +43,7 @@ export default class PreferenceExperimentAction extends Action {
             preferenceName,
             preferenceValue: branch.value,
             preferenceBranchType,
+            preferenceType,
```
CircleCI is failing partially for reasons unknown to me, but also because we need to update the tests for the action to assert that this parameter is passed as well.

Osmose commented on recipe-client-addon/test/unit/test_PreferenceExperiments.js:
```diff
@@ -165,6 +172,24 @@ add_task(withMockExperiments(withMockPreferences(async function (experiments, mo
+      name: "test",
+      branch: "branch",
+      preferenceName: "fake.type_preference",
+      preferenceBranchType: "invalid",
```
You'll want to use a valid preference branch type (`user`, in this case), or else `start` will throw for an invalid branch type, not due to an invalid preference type.

It looks like [`Assert.throws` takes an `expected` argument](https://dxr.mozilla.org/mozilla-central/source/testing/modules/Assert.jsm#381) that we can use to assert that error message matches the cause for the exception that we expect. If you follow the code it looks like we can pass a regex object in to match against a portion of the error message. Someday we should refactor the rest of the throws tests to use this too, it looks useful!

andymikulski commented:
I believe I've addressed all review feedback. I'm having tests fail reliably (again), even after rebasing onto `preference-experiment`. Really hoping I don't need to have to wipe my mozilla-central yet again, but I have a feeling I will.

Ready for feedback.

mythmon commented:
It looks like the only CI failures here are some lint problems in Circle, so we can probably merge this if it passes review.

If you get to the point where you need to wipe mozilla-central again, I'd like to have a vidyo call with you to see if we can work through the issue using less extreme methods. Maybe there is something easy we can figure out that won't require a full rebuild.

andymikulski commented:
Bummer. I thought I had fixed my eslint issue, but looks not. I'll correct now.
